### PR TITLE
bpo-35113: Fix inspect.getsource to return correct source for inner classes

### PR DESCRIPTION
**Pull request by tirkarthi**
Previous approach used regex along with indentation of the match to return the source code. This caused below issues. This PR proposes usage of ast module to do correct parsing.  

**Issues**

1. When a class with same name is defined under different classes.
2. When a class definition is part of a string declaration.
3. When a class definition is part of multiline comment.

Examples of incorrect output can be found with https://bugs.python.org/issue35113. 

This PR also adds returning class decorators as part of the source code instead of just returning the class definition. 

**Inner class example :** 

```python
import inspect

class Foo:
    class Bar:
        def foo(self):
            pass

class Spam:
    class Bar:
        def bar(self):
            pass

print(inspect.getsource(Spam.Bar))
```

**On master**

```
$ ./python.exe foo.py
    class Bar:
        def foo(self):
            pass

```

**IPython**

`inspect.getsource` is also used in lot of tools like IPython and pdb which also return incorrect output due to this

```python
ipython
Python 3.6.4 (default, Mar 12 2018, 13:42:53)
Type 'copyright', 'credits' or 'license' for more information
IPython 6.3.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import foo

In [2]: foo.Spam.Bar??
Init signature: foo.Spam.Bar()
Docstring:      <no docstring>
Source:
    class Bar:
        def foo(self):
            pass
File:           ~/stuff/python/backups/foo.py
Type:           type

$ python3 foo.py
> /Users/karthikeyansingaravelan/stuff/python/backups/foo.py(13)<module>()
-> if __name__ == "__main__":
(Pdb) pdb.getsourcelines(Spam.Bar)
(['    class Bar:\n', '        def foo(self):\n', '            pass\n'], 2)
```

**With PR**

```
$ ./python.exe foo.py
    class Bar:
        def bar(self):
            pass

```

**Class decorator example**

```python
import inspect

def foo(cls):
    return cls

@foo
@foo
class Foo:
    pass

print(inspect.getsource(Foo))
```

**On master**

```
$ ./python.exe foo.py
class Foo:
    pass

```

**With PR**

```
$ ./python.exe foo.py
@foo
@foo
class Foo:
    pass
```

https://bugs.python.org/issue35113

### CODE REVIEW

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: This decorator is only used once, and it is not general. I think the code would be simpler if just adds corresponding code in `visit_ClassDef()`.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Why only if the stack empty? What about deeper nesting?
```python
class A:
    class B:
        class C:
            pass
```

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Just look at the first item of `decorator_list` in `visit_ClassDef`.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Is it used?

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Are args and kwargs used?

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Just inline it.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Remove it and just add to the name to the stack at the beginning of `visit_ClassDef`. This will make the code simpler.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: I have refactored the code as per the comments. This was due to my stack management logic to add top level module and to push children class node inside the for loop. Since I have refactored the code to push and pop only at one place it's not needed. There is a similar test at https://github.com/python/cpython/pull/10307/files#diff-041789a617178e1a31ea7a1f86a5d2faR204 . Let me know if I need to add this also as a test.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Thanks, I have handled the decorator code inside `visit_ClassDef` with `node.decorator_list` and it's much simpler now.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Not needed. Removed. Thanks.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Not needed. Removed. Thanks.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: I feel this is more readable that I am joining the stack to form qualname than `if self.qualname == '.'join(self.stack)` . Let me know if it needs to be inlined.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Thanks, added it to the top and the function is much simpler to read than my original one :)

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: I think it would be clearer if remove empty lines in this method. Left empty lines only between methods and classes.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Since the class is not used outside of `findsource()` you can simplify the code by removing `__init__`. Inside the visitor you can just use local variables from the outer scope. This will save several lines. Add `nonlocal line_number` because it is set inside the visitor.

File: Lib/inspect.py
Comment (serhiy-storchaka) [reply Lib/inspect.py]: It is needed if use local variables from the outer scope. Then the condition will be `if qualname == '.'join(stack)`.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Just call `self.generic_visit(node)`.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Add the comment like in `_signature_fromstr()`.

**Comment by tirkarthi** [comment]
Thanks, I will work on them. It seems there is no test for similar to your examples since my PR raises OSError for your examples meanwhile master works. I guess my code breaks these cases which was handled by the previous regex based approach.

```python
def func123():
    class class234:
        def func(self):
            pass
    return class234

class class345:
    def func456(self):
        class class567:
            pass
        return class567

if __name__ == "__main__":
    import inspect
    print(inspect.getsource(func123()))
    print(inspect.getsource(class345().func456()))
```

```
$ python3 foo_1.py
    class class234:
        def func(self):
            pass

        class class567:
            pass

$ ../cpython/python.exe foo_1.py
Traceback (most recent call last):
  File "foo_1.py", line 15, in <module>
    print(inspect.getsource(func123()))
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/inspect.py", line 993, in getsource
    lines, lnum = getsourcelines(object)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/inspect.py", line 975, in getsourcelines
    lines, lnum = findsource(object)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/inspect.py", line 832, in findsource
    raise OSError('could not find class definition')
OSError: could not find class definition
```

I will fix the rest of the PR comments too in next commit.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Added a comment on lazy import of ast like `_signature_fromstr()`

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Changed. Thanks.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Removed empty lines. Thanks.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: I tried removing the constructor and using local variables which removes use of `self` as in `self.stack`. But since stack is changed in three functions now I have to declare `nonlocal` on three and make other changes. After conversion I think using class variables is more self-contained and I have kept them. Let me know if you need to change them to use local variables.

**Comment by tirkarthi** [comment]
One side effect of the change is that `inspect.getcomments` also works correctly. It caused a failure in test_pydoc at https://github.com/python/cpython/blob/master/Lib/test/test_pydoc.py#L479 . 

pydoc's `render_doc` adds some more content using `inspect.getdoc` or comments of a class using `inspect.getcomments` if either of them is present at https://github.com/python/cpython/blob/3e28eed9ec2249bb11ad0db4629271b7ce9b7918/Lib/pydoc.py#L94. Since on master the class prediction is wrong for two classes of same name the below on Python 3 returns None for `getcomments()` though there is a valid comment `# 124`. This causes the comment to be rendered on pydoc. In the linked `test_pydoc` function there was a comment which was also rendered along with the doc for `A()` causing test failure. So I have added an empty line between the `class A` and comment at 4d11679a9ab3997280a02ce6b5de37566d984444 to fix it.

```python3
import pydoc

def func123():
    # 124
    class class234:
        pass
    return class234

class class234:
    pass

if __name__ == "__main__":
    import inspect
    print(inspect.getcomments(func123()))
    print(pydoc.render_doc(func123()))
```

```
➜  backups python3 foo_3.py
None
Python Library Documentation: class class234 in module __main__

class class234(builtins.object)
 |  Data descriptors defined here:
 |
 |  __dict__
 |      dictionary for instance variables (if defined)
 |
 |  __weakref__
 |      list of weak references to the object (if defined)
```

```
➜  backups ../cpython/python.exe foo_3.py
# 124

Python Library Documentation: class class234 in module __main__

class class234(builtins.object)
 |  # 124
 |
 |  Data descriptors defined here:
 |
 |  __dict__
 |      dictionary for instance variables (if defined)
 |
 |  __weakref__
 |      list of weak references to the object (if defined)
```

Seems there is an altered environment failure with my test I will look into it. I think I am running the asyncdef version of the test correctly at https://github.com/python/cpython/pull/10307/files#diff-0258c7716015b67c231e0d2a60ebbd31R711. Please add in your thoughts.

Thanks!

File: Lib/inspect.py
Comment (serhiy-storchaka) [reply Lib/inspect.py]: No, you don't have to use `nonlocal` with a stack, because you don't assign to this variable in these methods. Only `line_number` needs it, at only in a single place. I prefer more compact code, but it is up to you using the construct and instance attributes.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Just `visit_AsyncFunctionDef = visit_FunctionDef`.

File: Lib/test/inspect_fodder2.py
Comment (serhiy-storchaka) [inline Lib/test/inspect_fodder2.py]: Maybe use `cls205` for testing classes with same name but different `__qualname__`?

File: Lib/test/inspect_fodder2.py
Comment (serhiy-storchaka) [inline Lib/test/inspect_fodder2.py]: This method is not needed.

File: Lib/test/inspect_fodder2.py
Comment (serhiy-storchaka) [inline Lib/test/inspect_fodder2.py]: This method is not needed.

File: Lib/test/test_inspect.py
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: Use `addCleanup()`.

File: Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst]: Add "Path by ...".

File: Lib/test/test_inspect.py
Comment (tirkarthi) [reply Lib/test/test_inspect.py]: Seems this test class inherits from `unittest.Test` and `self.addCleanup(asyncio.set_event_loop_policy, None)` doesn't work. Should I use `asyncio.set_event_loop_policy(None)` or change the base class from which it's inherited?

Comment (tirkarthi) [reply Lib/test/test_inspect.py]: I tried adding `class TestBuggyCases(GetSourceBase, test_utils.TestCase):` but it seems it expects all the test cases to setup an event loop and gives below error. Any other better way to do this?

```
======================================================================
ERROR: test_with_comment_instead_of_docstring (test.test_inspect.TestBuggyCases)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_asyncio/utils.py", line 533, in tearDown
    self.unpatch_get_running_loop()
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_asyncio/utils.py", line 525, in unpatch_get_running_loop
    events._get_running_loop = self._get_running_loop
AttributeError: 'TestBuggyCases' object has no attribute '_get_running_loop'
```

Comment (serhiy-storchaka) [reply Lib/test/test_inspect.py]: Why it doesn't work? There is no `unittest.Test`, this class inherits from `unittest.TestCase`.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Thanks, I will add this. I will also refactor class to use local variables. My assumption was that `stack.append` requires stack variable to be nonlocal since it mutates `stack`. I was wrong on that part. Thanks for clearing up and I will change it as suggested to use variables from outer scope making it more compact 👍

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: Not needed.

File: Lib/test/test_inspect.py
Comment (serhiy-storchaka) [inline Lib/test/test_inspect.py]: It should be called before the first `asyncio.run()`. The purpose of using `addCleanup` is that the clean up code should be executed if any code above fails.

File: Lib/test/test_inspect.py
Comment (tirkarthi) [reply Lib/test/test_inspect.py]: Sorry, it works. I think I was having test failures and tested it during that time causing confusion that it's an environment change. Added the change as part of c6e5e2c133069bcb45568b141636b6b78e2348d3 and verified locally.

```
./python.exe -m test --fail-env-changed test_inspect
Run tests sequentially
0:00:00 load avg: 2.57 [1/1] test_inspect

== Tests result: SUCCESS ==

1 test OK.

Total duration: 2 sec 345 ms
Tests result: SUCCESS
```

File: Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst
Comment (tirkarthi) [reply Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst]: Added patch attribute. Thanks :)

File: Lib/test/inspect_fodder2.py
Comment (tirkarthi) [reply Lib/test/inspect_fodder2.py]: Removed. Thanks

File: Lib/test/inspect_fodder2.py
Comment (tirkarthi) [reply Lib/test/inspect_fodder2.py]: Removed. Thanks

File: Lib/test/inspect_fodder2.py
Comment (tirkarthi) [reply Lib/test/inspect_fodder2.py]: Used `cls205` in the test. Thanks.

File: Lib/inspect.py
Comment (serhiy-storchaka) [inline Lib/inspect.py]: `ClassVisitor` is not very informative name. It visits not only classes. Maybe name it `ClassFinder` or something like?

File: Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst]:
```suggestion
with same name as module level class. Class decorators are also returned as
```

**Review by serhiy-storchaka — COMMENTED** [review]
Just few nitpicks.

But see also the comment on the tracker.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Renamed to `ClassFinder` and the corresponding variable below to `class_finder`. Thanks.

File: Lib/test/test_inspect.py
Comment (tirkarthi) [reply Lib/test/test_inspect.py]: Thanks I moved it up before first `asyncio.run` and tested the same.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: Removed. Thanks.

File: Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst
Comment (tirkarthi) [reply Misc/NEWS.d/next/Library/2018-11-03-16-18-20.bpo-35113.vwvWKG.rst]: Changed. Thanks.

File: Lib/test/inspect_fodder2.py
Comment (matrixise) [inline Lib/test/inspect_fodder2.py:218]: should be `cls218`

File: Lib/test/inspect_fodder2.py
Comment (matrixise) [inline Lib/test/inspect_fodder2.py:231]: should be `cls231`

File: Lib/inspect.py
Comment (matrixise) [inline Lib/inspect.py]: why not "line_number -= 1" ?

**Review by matrixise — CHANGES_REQUESTED** [review]
thanks for your contributions

File: Lib/test/inspect_fodder2.py
Comment (tirkarthi) [reply Lib/test/inspect_fodder2.py:218]: This to check that the inner class which has the same name `cls213` inside `func212` doesn't conflict with this class `cls213`. Previously regex was used and hence there were cases where two classes with same name defined under different scopes might conflict with each other returning wrong results.

File: Lib/test/inspect_fodder2.py
Comment (tirkarthi) [reply Lib/test/inspect_fodder2.py:231]: Same as my above comment but this is to check for `async def` related code.

This to check that the inner class which has the same name `cls226` inside `func225` doesn't conflict with this class `cls226`. Previously regex was used and hence there were cases where two classes with same name defined under different scopes might conflict with each other returning wrong results.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: No specific reason to be honest :) I will update it along with other review comments in the second round of review. Thanks.

**Comment by serhiy-storchaka** [comment]
I don't think a deque has advantages over a list here. We push and pop from one end.

Stopping when the first candidate is found will make finding `cls1999` in your example 10x faster. But it will not help in case of `cls19999`.

Yes, parsing to AST is much slower than using a regex, and uses much more memory, but I don't think this is very large issue. In common cases the performance should be good enough.

**Comment by tirkarthi** [comment]
@1st1 It would be great if you can review this

**Comment by 1st1** [comment]
Looks good to me in general.  Here're a few questions:

1. For the example from bpo:
    ```python
    if sys.platform == 'win32':
        class Arena(object):
            ...
    else:
        class Arena(object):
            ...
    ```

   Which class will be returned?  Serhiy mentioned that the current implementation returns the first one.  Did you fix this implementation to do the same?

2. > This PR also adds returning class decorators as part of the source code instead of just returning the class definition.

    This might be a backwards compatibility issue.  I'd prefer the behaviour to stay absolutely the same.  We can add another function, or can consider adding a keyword-only flag to `getsource` to make it return decorators.

**Comment by 1st1** [comment]
> I guess @serhiy-storchaka brought this up since functions return decorators as part of inspect.getsource. I have modified the PR and NEWS entry to make sure decorators are not returned like the original implementation. I will open a separate enhancement request on bpo for this.

It's an interesting discrepancy.  For the sake of making the API consistent we might consider breaking it slightly for classes and start returning decorators too in 3.8. I honestly don't think it's going to be a huge deal. I'm curious what @serhiy-storchaka thinks about that though.

**Comment by serhiy-storchaka** [comment]
I think it is good to start returning decorators too.

**Comment by tirkarthi** [comment]
Agreed with @serhiy-storchaka that this is a useful enhancement and makes the API consistent with functions. Since the decorator related code is reverted @1st1 please let me know on your decision so that I can add it back if needed.

Thanks

**Comment by tirkarthi** [comment]
Adding decorators to classes too was an enhancement proposed by https://bugs.python.org/issue15856 . I have currently reverted the decorator related code but I think it would be good to add it back to return decorators for consistency.

**Comment by tirkarthi** [comment]
@1st1 It will be helpful if you can review the changes. I have changed the code to return decorators too to resolve https://bugs.python.org/issue15856 and also it makes it consistent with source code for functions. Not sure if this can be merged after beta release given the decorator enhancement but I would be very happy if this makes it part of 3.8 beta 1.

Thanks

File: Lib/inspect.py
Comment (1st1) [inline Lib/inspect.py]: I don't like this way of interrupting the AST traversal.  Just declare a custom exception, raise it, and then catch it.

File: Lib/inspect.py
Comment (1st1) [inline Lib/inspect.py]: I think you should move this class to the top level, and instead of using a nonlocal name, use a proper property, e.g. `self.line_number`.

File: Lib/inspect.py
Comment (1st1) [inline Lib/inspect.py]: I think you can just import it at the top level. The `inspect` module is a pretty heavy module already.

**Review by 1st1 — CHANGES_REQUESTED** [review]
Almost there. I've left a few nit comments.

File: Lib/inspect.py
Comment (tirkarthi) [reply Lib/inspect.py]: I am currently using the name `_ClassFinder` similar to (_ParameterKind) and `ClassFoundException` respectively. Please let me know if I need to change these names or move the location of these definitions in the file.

Thanks

**Comment by tirkarthi** [comment]
I have made the requested changes; please review again

* Moved import ast to top.
* Used custom exception for to raise exception with line_number instead of StopIteration.
* Move class finder code as a class with attributes instead of using nonlocal variables.

**Comment by tirkarthi** [comment]
Not sure why Windows test_json fails with `RecursionError` over addressing the PR comments which is not reproducible locally. I don't have a windows machine to debug this and rebasing to master also doesn't help :( json module or tests doesn't seem to import anything from inspect module.

**Comment by tirkarthi** [comment]
@pablogsal This PR caused some errors on Appveyor regarding json tests. Appveyor was removed but I would still like to have a custom build of this just to be sure. Does adding the label to trigger custom builds okay for this?

**Comment by pablogsal** [comment]
Absolutely!

Remember to inspect any possible failure individually as the buildbots can fail due to other unrelated problems (the same problems that are currently failing on master).

**Comment by tirkarthi** [comment]
Thanks @pablogsal I have rebased since the PR is old and will trigger a custom buildbot test once the primary CI passes.

**Comment by serhiy-storchaka** [comment]
Oh, it is not merged yet! What problems are left?

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by tirkarthi** [comment]
There were some weird problems with test_json on Windows which I assume is due to importing ast module on the top but sounds more unrelated.

**Comment by tirkarthi** [comment]
@serhiy-storchaka I have rebased the branch with latest master since it was old. It will be helpful to have your review and re-approval so that I can merge this.

**Review by serhiy-storchaka — APPROVED** [review]
:+1:

I just have found a way to optimize the visitor, but it may complicate the implementation. Will try after merging this PR.

**Comment by tirkarthi** [comment]
Thanks Serhiy, I have merged the PR. Unfortunately, I had an issue with GitHub UI that my squashed message in the textbox didn't pass through and it just added the intermediate commit messages even after edit :(

**Comment by carljm** [comment]
FWIW the performance regression due to this PR made `inspect.getsource` unusable for us in an existing use case in a large code base. Just noting it here rather than filing an issue since it seems like this was already considered here and it was decided that performance of `inspect.getsource` is not important. And I'm not sure it's really fixable anyway without reintroducing the bugs fixed by this PR, and it's too late for that. Just making a note for historical value to clarify that this scale of performance regression will likely bite someone.

**Pull request merged**
